### DISCUSSION TITLE
Support jsonpath-like column select syntax in the expression language

### DISCUSSION THREAD

**Discussion by matthewgdv** [sqlalchemy/sqlalchemy#10786] [Ideas]
### Describe the use case

Redshift provides a `SUPER` data type that is intended to hold unstructured data. This is similar to the `JSON` type in postgres, except that it's a more generic approach. For instance, while a common way to create a `SUPER` value is to use Redshift's built-in `json_parse` function, there's nothing stopping users from defining functions to parse other serialization formats into `SUPER`.

Traversing a `SUPER` value uses a SQL-level syntax (i.e. not within a string) that is very similar to a jsonpath. I've provided a toy example further down to show off how it works.

### Databases / Backends / Drivers targeted

Redshift has had this syntax for its new `SUPER` type since 2022. I don't know enough about other backends to know if other dialects also have similar syntax.

Regardless, there would be no reason to make an implementation for this dialect-specific. Imo it would make more sense to provide this as a method on the `Column` class (or an appropriate ancestor).

### Example Use

```SQL
SELECT
    sub.data[0].name::varchar AS first_dog_name,
    sub.data[0].age::int AS first_dog_age,
    sub.data[0].breed.name::varchar AS first_dog_breed,
    sub.data[1].name::varchar AS second_dog_name,
    sub.data[1].age::int AS second_dog_age
FROM (
    SELECT
        json_parse('[{"name": "Luna", "age": 2, "breed": {"name": "Golden Retriever"}}, {"name": "Cooper", "age": 5}]') as data
) as sub;
```

![image](https://github.com/sqlalchemy/sqlalchemy/assets/37582022/f6353c7b-d528-4ef7-904c-fb1d36d9ce9c)

### Additional context

Implementing this should be relatively straightforward using python's `__getattr__` and `__getitem__` dunders. In a way it's very similar to what the `func` accessor does, where it just keeps on giving proxy objects that 'remember' the full dotted path that's been traversed so far.

I'm not sure what a good name for the initial property that lets you access the first proxy column object would be, but just for the sake of offering a bikesheddable example, let's use the property name `.traverse`, so maybe something like this:

```python
select(
    cast(sub.data.traverse[0].name, String).label('first_dog_name'),
    cast(sub.data.traverse[0].age, Integer).label('first_dog_age'),
    cast(sub.data.traverse[0].breed.name, String).label('first_dog_breed'),
    ...
)
```

**Comment by CaselIT** [comment]
Hi,

I don't think there is anything to implement on sqlalchemy side, since this all seems supported.
You would have to ask the redshift driver developers to support the super type in a similar way to what the json type already allows.

For example:
```py
import sqlalchemy as sa
import sqlalchemy.dialects.postgresql as pg

x = sa.func.json_parse(
    '[{"name": "Luna", "age": 2, "breed": {"name": "Golden Retriever"}}, {"name": "Cooper", "age": 5}]'
).table_valued(sa.column("data", pg.JSONB)).alias('sub')

q = sa.select(
    x.c.data[0]['name'].label('first_dog_name'),
    x.c.data[0]['age'].as_integer().label('first_dog_age'),
    x.c.data[0]['breed']['name'].label('first_dog_breed'),
)
print(q.compile(dialect=pg.dialect(), compile_kwargs={'literal_binds': True}))
```
this prints
```sql
SELECT (sub.data -> 0) -> 'name' AS first_dog_name, CAST((sub.data -> 0) ->> 'age' AS INTEGER) AS first_dog_age, ((sub.data -> 0) -> 'breed') -> 'name' AS first_dog_breed
FROM json_parse('[{"name": "Luna", "age": 2, "breed": {"name": "Golden Retriever"}}, {"name": "Cooper", "age": 5}]') AS sub
```

**Reply by CaselIT** [reply]
btw the query above does not work on pg, but this does
```sql
SELECT (sub.data -> 0) -> 'name' AS first_dog_name, CAST((sub.data -> 0) ->> 'age' AS INTEGER) AS first_dog_age, ((sub.data -> 0) -> 'breed') -> 'name' AS first_dog_breed
FROM (select '[{"name": "Luna", "age": 2, "breed": {"name": "Golden Retriever"}}, {"name": "Cooper", "age": 5}]'::jsonb AS data) sub
```

generated with
```py
import sqlalchemy as sa
import sqlalchemy.dialects.postgresql as pg

x = sa.select(sa.literal(
    '[{"name": "Luna", "age": 2, "breed": {"name": "Golden Retriever"}}, {"name": "Cooper", "age": 5}]'
).cast(pg.JSONB).label('data')).subquery('sub')

q = sa.select(
    x.c.data[0]['name'].label('first_dog_name'),
    x.c.data[0]['age'].as_integer().label('first_dog_age'),
    x.c.data[0]['breed']['name'].label('first_dog_breed'),
)
print(q.compile(dialect=pg.dialect(), compile_kwargs={'literal_binds': True}))
```

overall once SUPER supports the functionality in json and uses the proper compile syntax everything should be supported. Both of these are dialect changes though

**Reply by matthewgdv** [reply]
Wow, I wasn't aware `sqlalchemy` columns already implemented `__getitem__` like this. My bad then, sorry for the bother. At least having this discussion up online might provide something for the next person in my shoes to google. Out of curiosity, is it documented somewhere? I assume it is, but I just didn't know the right place to look or what keywords to use.

Thanks for the quick response :)

**Reply by CaselIT** [reply]
only json and jsonb columns have it, but it's something can should be easily adaptable to other types. the main thing to change would be how you render the parts in sql.

I think the best docs are here https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.JSON
